src/relo_core.c

```diff
@@ -1234,55 +1234,68 @@ static void btf_reloc_append_type(struct btf_reloc_info *info, struct btf_reloc_
 
 // add type resolving all typedefs
 static void btf_reloc_add_type(struct btf *btf, struct btf_reloc_info *info, struct btf_reloc_type *reloc_type) {
-	// append this type to the list
+
+	// append this type to the relocation type's list before anything else
 	btf_reloc_append_type(info, reloc_type);
 
-	unsigned int id = reloc_type->type->type;
+	struct btf_array *array;
+	struct btf_type *btf_type;
+	struct btf_reloc_type *current_type;
 
-	// resolve typedefs
-	if (btf_is_typedef(reloc_type->type)) {
-		struct btf_type *btf_type = (struct btf_type *) btf__type_by_id(btf, id);
+	unsigned int id = reloc_type->type->type;
 
+	switch (btf_kind(reloc_type->type)) {
+	case BTF_KIND_INT:
+	case BTF_KIND_PTR:
+	case BTF_KIND_STRUCT:
+	case BTF_KIND_UNION:
+	case BTF_KIND_VOLATILE:
+		break;
+	case BTF_KIND_TYPEDEF:
```

Comment:
```suggestion
	case BTF_KIND_UNION:
		break;
	case BTF_KIND_VOLATILE:
	case BTF_KIND_TYPEDEF:
```

Otherwise the volatile's type won't be added on the generated BTF.